codegen/src/main/scala/overflowdb/codegen/CodeGen.scala

```diff
@@ -1556,34 +1556,54 @@ class CodeGen(schema: Schema) {
          |}
          |""".stripMargin
 
-    def generateNewNodeSource(nodeType: NodeType, keys: Seq[Property[_]]) = {
+    def generateNewNodeSource(nodeType: NodeType, properties: Seq[Property[_]]) = {
       import Property.Cardinality
-      val fieldDescriptions = mutable.ArrayBuffer.empty[(String, String, String)] // fieldName, type, default
-      for (key <- keys) {
-        val defaultImpl = key.cardinality match {
-          case Cardinality.ZeroOrOne => "None"
-          case Cardinality.List => "collection.immutable.ArraySeq.empty"
-          case Cardinality.One(default) => defaultValueImpl(default)
-        }
-        val typ = getCompleteType(key)
-        fieldDescriptions += ((camelCase(key.name), typ, defaultImpl))
+      case class FieldDescription(name: String, valueType: String, fullType: String, cardinality: Cardinality)
+      val fieldDescriptions = mutable.ArrayBuffer.empty[FieldDescription]
+
+      for (property <- properties) {
+        fieldDescriptions += FieldDescription(
+          camelCase(property.name),
+          typeFor(property),
+          getCompleteType(property),
+          property.cardinality)
       }
+
       for (containedNode <- nodeType.containedNodes) {
-        val defaultImpl = containedNode.cardinality match {
-          case Cardinality.ZeroOrOne => "None"
-          case Cardinality.List => "collection.immutable.ArraySeq.empty"
-          case Cardinality.One(default) => defaultValueImpl(default)
-        }
-        val typ = getCompleteType(containedNode)
-        fieldDescriptions += ((containedNode.localName, typ, defaultImpl))
+        fieldDescriptions += FieldDescription(
+          containedNode.localName,
+          typeFor(containedNode),
+          getCompleteType(containedNode),
+          containedNode.cardinality)
       }
-      val defaultsVal = fieldDescriptions.reverse
-        .map {
-          case (name, typ, default) => s"var $name: $typ = $default"
-        }.mkString(", ")
+
+      val memberVariables = fieldDescriptions.reverse.map {
+        case FieldDescription(name, _, fullType, cardinality) =>
+          val defaultValue = cardinality match {
+            case Cardinality.One(default) => defaultValueImpl(default)
+            case Cardinality.ZeroOrOne => "None"
+            case Cardinality.List => "collection.immutable.ArraySeq.empty"
+          }
+          s"var $name: $fullType = $defaultValue"
+
+      }.mkString(", ")
+
+      val builderSetters = fieldDescriptions.map {
+        case FieldDescription(name, valueType, _, cardinality) =>
+          val (parameterTypeX, assignParameterX) = cardinality match {
+            case Cardinality.One(_) => (valueType, "x")
+            case Cardinality.ZeroOrOne => (valueType, "Option(x)")
+            case Cardinality.List => (s"IterableOnce[$valueType]", "x.to(collection.immutable.ArraySeq)")
+          }
```

Comment:
Theoretically we could just make an overload, right?

I.e. optional parameters can be set with either `builder.foo("bar")` or with `builder.foo(Some("bar"))`?

Otherwise, we may get a compat problem. 

It would be especially nice if `builder.lineNumber(17)`, `builder.lineNumber(Integer.valueOf(17))` and `builder.lineNumber(None)` and `builder.lineNumber(Some(17))` all worked.

Reply:
as discussed via zoom, i just brought those back in as deprecated overloads, mostly to have a happy migration path for frontends